File: JS/lib/linked_list.test.js
import assert from "node:assert";
import { describe, it } from "node:test";
import { ListNode } from "./linked_list.js";

describe("ListNode", () => {
	it("should create a ListNode with the correct value and next node", () => {
		const node = new ListNode(5);
		assert.strictEqual(node.val, 5);
		assert.strictEqual(node.next, null);
	});

	it("should create a ListNode with a given value and next node", () => {
		const node2 = new ListNode(10);
		const node1 = new ListNode(5, node2);
		assert.strictEqual(node1.val, 5);
		assert.strictEqual(node1.next, node2);
		assert.strictEqual(node2.val, 10);
		assert.strictEqual(node2.next, null);
	});

	it('should create a linked list from an array of numbers using the "of" method', () => {
		const list = ListNode.of(1, 2, 3, 4, 5);
		assert.strictEqual(list.val, 1);
		assert.strictEqual(list.next.val, 2);
		assert.strictEqual(list.next.next.val, 3);
		assert.strictEqual(list.next.next.next.val, 4);
		assert.strictEqual(list.next.next.next.next.val, 5);
		assert.strictEqual(list.next.next.next.next.next, null);
	});

	it('should return null when calling "of" with an empty array', () => {
		const list = ListNode.of();
		assert.strictEqual(list, null);
	});

	it("return string representation of linked list correctly", () => {
		const list = ListNode.of(1, 2, 3, 4).toString();
		assert.strictEqual(list, "1 -> 2 -> 3 -> 4");
	});
});

File: src/lib/linked_list_test.cc
#include "linked_list.hpp"
#include <gtest/gtest.h>
#include <string>
#include <vector>

class LinkedListTest : public testing::TestWithParam<std::vector<int>> {};

// --- Test: Create List ---
TEST_P(LinkedListTest, CreateList) {
  auto list = make_list(GetParam());

  const auto &param = GetParam();

  ListNode *current = list;
  for (size_t i = 0; i < param.size(); ++i) {
    ASSERT_NE(current, nullptr);
    EXPECT_EQ(current->val, param[i]);
    current = current->next;
  }
  EXPECT_EQ(current, nullptr); // End of list

  delete_list({list});
}

// --- Test: Check equality between lists ---
TEST_P(LinkedListTest, Equal) {
  auto listA = make_list(GetParam());
  auto listB = make_list(GetParam());

  EXPECT_TRUE(*listA == *listB);

  delete_list({listA, listB});
}

// --- Test: Check inequality between lists ---
TEST_P(LinkedListTest, InEqual) {
  auto listA = make_list(GetParam());
  auto listB = make_list(GetParam());

  // Modify listB to make it different from listA
  if (listB->next) {
    listB->next->val += 1; // Change the value of the second node
  }

  EXPECT_FALSE(*listA == *listB);

  delete_list({listA, listB});
}

// --- Test: to_string function ---
TEST_P(LinkedListTest, ToString) {
  auto list = make_list(GetParam());
  std::string result = list->to_string();

  const auto &param = GetParam();
  std::string expected = "";
  for (size_t i = 0; i < param.size(); ++i) {
    expected += std::to_string(param[i]);
    if (i < param.size() - 1) {
      expected += " -> ";
    }
  }

  EXPECT_EQ(result, expected);

  delete_list({list});
}

INSTANTIATE_TEST_SUITE_P(, LinkedListTest,
                         testing::Values(std::vector<int>{1, 2, 3},
                                         std::vector<int>{4, 5, 6},
                                         std::vector<int>{10, 20, 30},
                                         std::vector<int>{99, 100, 101},
                                         std::vector<int>{1, 2, 3},
                                         std::vector<int>{10, 20, 30}));
